db/repos/accounts.js

```diff
@@ -25,34 +30,311 @@ var columnSet;
-		'secondPublicKey',
-		'balance',
-		'rate',
-		'rank'
-	];
-
-	if (!columnSet) {
-		columnSet = {};
-		var table = new pgp.helpers.TableName({table: this.dbTable, schema: 'public'});
-		columnSet.insert = new pgp.helpers.ColumnSet(this.dbFields, {table: table});
+class AccountsRepo {
+
+	constructor (db, pgp) {
+		this.db = db;
+		this.pgp = pgp;
+
+		this.dbTable = 'mem_accounts';
+
+		const ifNotExists = c => !c.exists;
+
+		// Used in SELECT, UPDATE, INSERT
+		const normalFields = [
+			{name: 'isDelegate', cast: 'int::boolean', skip: ifNotExists},
+			{name: 'u_isDelegate', cast: 'int::boolean', skip: ifNotExists},
+			{name: 'secondSignature', cast: 'int::boolean', skip: ifNotExists},
+			{name: 'u_secondSignature', cast: 'int::boolean', skip: ifNotExists},
+			{name: 'balance', cast: 'bigint', def: '0', skip: ifNotExists},
+			{name: 'u_balance', cast: 'bigint', def: '0', skip: ifNotExists},
+			{name: 'rate', cast: 'bigint', def: '0'},
+			{name: 'multimin', def: 0},
+			{name: 'u_multimin', def: 0},
+			{name: 'multilifetime', def: 0},
+			{name: 'u_multilifetime', def: 0},
+			{name: 'blockId', def: null},
+			{name: 'nameexist', def: 0},
+			{name: 'u_nameexist', def: 0},
+			{name: 'fees', cast: 'bigint', def: '0'},
+			{name: 'rewards', cast: 'bigint', def: '0'},
+			{name: 'vote', cast: 'bigint', def: '0'},
+			{name: 'producedblocks', cast: 'bigint', def: '0', prop: 'producedBlocks'},
+			{name: 'missedblocks', cast: 'bigint', def: '0', prop: 'missedBlocks'},
+			{name: 'username', def: null, skip: ifNotExists},
+			{name: 'u_username', def: null, skip: ifNotExists},
+			{name: 'publicKey', mod: ':raw', init: () => 'ENCODE("publicKey", \'hex\')', skip: ifNotExists},
+			{name: 'secondPublicKey', mod: ':raw', init: () => 'ENCODE("secondPublicKey", \'hex\')', skip: ifNotExists},
+			{name: 'virgin', cast: 'int::boolean', def: 1, skip: ifNotExists}
+		];
+
+		// ONLY USED IN SELECT and INSERT
+		const immutableFields = [
+			{ name: 'address', mod: ':raw', init: () => 'UPPER("address")'}
+		];
+
+		// ONLY USED IN SELECT
+		const dynamicFields = [
+			{name: 'rank', cast: 'bigint', mod: ':raw', init: () => 'row_number() OVER (ORDER BY "vote" DESC, "publicKey" ASC)'},
+			{name: 'delegates', mod: ':raw', init: () => '(SELECT ARRAY_AGG("dependentId") FROM mem_accounts2delegates WHERE "accountId" = "mem_accounts"."address")'},
+			{name: 'u_delegates', mod: ':raw', init: () => '(SELECT ARRAY_AGG("dependentId") FROM mem_accounts2u_delegates WHERE "accountId" = "mem_accounts"."address")'},
+			{name: 'multisignatures', mod: ':raw', init: () => '(SELECT ARRAY_AGG("dependentId") FROM mem_accounts2multisignatures WHERE "accountId" = "mem_accounts"."address")'},
+			{name: 'u_multisignatures', mod: ':raw', init: () => '(SELECT ARRAY_AGG("dependentId") FROM mem_accounts2u_multisignatures WHERE "accountId" = "mem_accounts"."address")'}
+		];
+
+		this.dbFields = _.union(normalFields, immutableFields, dynamicFields);
+
+		if (!columnSet) {
```

Comment:
This should be just above line 41. No point declaring and merging columns outside of this verification.